Type-check stubs

randycoulman opened the pull request:
Part of #56

Currently, only mocks created with `expect` are type-checked by Hammox.

This provides the same protection for stubs created with `stub`.

There is also `stub_with`, but I'm not sure how to add protection for that one; suggestions welcome!

I wasn't sure how best to test this change. For now, I modified `assert_pass` and `assert_fail` to use `stub` instead of `expect`; there are a few other tests that continue to use `expect`, so both code paths are being tested.  I'm happy to change this up however you'd like.

I extracted a helper named `wrap` that is now used in both `expect` and `stub`; I'd rather use the name `protected` for this, but that name's already taken. Again, I'm happy to change this however you'd like.

msz commented:
Thanks! I think the code looks great.

For testing, I'd go a simpler way — I'd just make a test where we stub with a wrong function and try to call the stub, then assert that there was a Hammox exception thrown. This should ensure that stubs are being protected.

randycoulman commented:
@msz Great, thanks! I'll take care of that tomorrow.

If you have a minute to drop a `hacktoberfest-accepted` label on this PR, I'd appreciate it.  Thanks!

randycoulman commented:
@msz I've updated the tests as you suggested; I reverted my earlier change and added explicit tests of `expect/4` and `stub/3` near the bottom of the test file.

msz reviewed (approved):
Looks great, thanks!
